### PR TITLE
introduce remediation

### PR DESCRIPTION
**Pull request by enxebre**
introduce remediation logic:
- if machine belongs to healthCheck CRD then remediate
- Remediate:
   - if it is master skip
   - if is not owned by a machineSet skip
   - if it is recently unhealthy requeue
   - if it is unhealthy for too long delete

### CODE REVIEW

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 1]
```diff
@@ -140,18 +146,93 @@ func getMachineHealthCheckListOptions() *client.ListOptions {
...
+	if !hasMachineSetOwner(*machine) {
+		glog.Info("Machine %s has no machineSet controller owner, skipping remediation", machine.Name)
+		return reconcile.Result{}, err
+	}
+
 	if !isHealthy(node) {
+		if unhealthyForTooLong(node) {
+			// delete
+			glog.Infof("machine %s has been unhealthy for too long, deleting", machine.Name)
+			if err := r.client.Delete(context.TODO(), machine); err != nil {
```
[the remark this answers is not in the document]
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: We are adding support for that at the actuator level so no api consumer can delete masters. I'll make sure this is also implemented and unit tested here

[the remark this answers is not in the document]
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: added https://github.com/openshift/machine-api-operator/pull/118/commits/771e6f416339c7f556085a82f9dd1e8bd1e1636f

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -120,14 +127,15 @@ func (r *ReconcileMachineHealthCheck) Reconcile(request reconcile.Request) (reco
 	for _, hc := range allMachineHealthChecks.Items {
 		if hasMatchingLabels(&hc, machine) {
 			glog.V(4).Infof("Machine %s has a matching machineHealthCheck: %s", machineKey, hc.Name)
-			remediate(node)
+			return remediate(r, machine)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: We're deliberating ignoring the rest of `range allMachineHealthChecks.Items` - yes? if so, why don't the other items come into play?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 3]
```diff
@@ -265,19 +284,193 @@ func TestReconcile(t *testing.T) {
 		}
 		result, err := r.Reconcile(request)
 		if result != tc.expected.result {
-			t.Errorf("Expected %v, got: %v", tc.expected.result, result)
+			t.Errorf("Test case : %v. Expected: %v, got: %v", tc.node.Name, tc.expected.result, result)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: nit: `"Test case:` (i.e., no space) like the formatting in `TestGetNodeCondition`.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 4]
```diff
@@ -265,19 +284,193 @@ func TestReconcile(t *testing.T) {
...
 		}
 		if tc.expected.error != (err != nil) {
-			t.Errorf("Expected error, got %v", err)
+			var errorExpectation string
+			if !tc.expected.error == true {
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Drop `== true` and just use as the implicit boolean expression.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 5]
```diff
@@ -265,19 +284,193 @@ func TestReconcile(t *testing.T) {
...
-func newFakeReconciler(initObjects ...runtime.Object) reconcile.Reconciler {
+func newFakeReconciler(initObjects ...runtime.Object) *ReconcileMachineHealthCheck {
 	fakeClient := fake.NewFakeClient(initObjects...)
 	return &ReconcileMachineHealthCheck{
 		client: fakeClient,
 		scheme: scheme.Scheme,
 	}
 }
+
+func TestHasMachineSetOwner(t *testing.T) {
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: nit: wording: `TestMachineSetHasOwner`?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 6]
```diff
@@ -265,19 +284,193 @@ func TestReconcile(t *testing.T) {
...
+
+	testsCases := []struct {
+		machine  *capiv1alpha1.Machine
+		expected expectedReconcile
+	}{
+		{
+			machine: machineWithNodeHealthy,
+			expected: expectedReconcile{
+				result: reconcile.Result{},
+				error:  false,
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: I would advocate using the "zero" value here, and for the other cases that default to `false`. Omitting this entry will default to `false`.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 7]
```diff
@@ -261,3 +265,33 @@ func hasMatchingLabels(machineHealthCheck *healthcheckingv1alpha1.MachineHealthC
...
+	machineLabels := labels.Set(machine.Labels)
+	for _, masterLabel := range machineMasterLabels {
+		if machineLabels.Get(masterLabel) == "master" {
+			return true
+		}
+	}
+
+	node, err := getNodeFromMachine(machine, client)
+	if err != nil {
+		glog.Warningf("Could't get node for machine %s", machine.Name)
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: typo: "Couldn't"

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go [hunk 8]
```diff
@@ -474,3 +475,57 @@ func TestRemediate(t *testing.T) {
...
+		machine  *capiv1alpha1.Machine
+		expected bool
+	}{
+		{
+			machine:  masterMachine,
+			expected: true,
+		},
+		{
+			machine:  workerMachine,
+			expected: false,
```
Comment (frobware) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Can drop this; `false` is the "zero" value.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 5 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: Actually is right, is testing if "it has a machineSet owner"? happy to reword if you have a better suggestion

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go (hunk 2 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: The initial approach is simple. A MachineHealthCheck CRD is not configurable, it just select machines based on labels, if there's a least one matching a machine it should be remediated.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 5 above)
Comment (frobware) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: ok

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go (hunk 7 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: thanks

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 6 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: I'd rather keep it explicitly for the test cases, it makes it easier to visualise and understand the expectation when looking at it

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go (hunk 8 above)
Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go]: thanks thanks

**Pull request merged**